src/realm/object-store/impl/results_notifier.cpp

```diff
@@ -164,8 +145,27 @@ void ResultsNotifier::run()
         return;
     }
 
-    if (!need_to_run())
+    {
+        auto lock = lock_target();
+        // Don't run the query if the results aren't actually going to be used
+        if (!get_realm() || (!have_callbacks() && !m_results_were_used))
+            return;
+    }
+
+    if (has_run() && m_query->sync_view_if_needed() == m_last_seen_version) {
+        // We've run previously and none of the tables involved in the query
+        // changed so we don't need to rerun the query, but we still need to
+        // check each object in the results to see if it was modified
+        if (!any_related_table_was_modified(*m_info))
+            return;
+        auto checker = get_modification_checker(*m_info, m_query->get_table());
+        for (size_t i = 0; i < m_previous_rows.size(); ++i) {
+            if (checker(m_previous_rows[i])) {
```

Comment:
Maybe faster for some results that have a deep change checker, at the expense of possibly running `IndexSet:::find(i)` twice if it doesn't exist and needs to be added?
```suggestion
            if (!m_change.modifications.contains(i) && checker(m_previous_rows[i])) {
```

Reply:
`m_change` has to be empty at this point. If it was non-empty then that means that the query's table was modified and we'd be rerunning the query.

Reply:
Ok that makes sense, thanks for explaining!